### DISCUSSION TITLE
0049 XLS-49d: Multiple Signer Lists

### DISCUSSION THREAD

**Discussion by mvadari** [XRPLF/XRPL-Standards#144] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by shortthefomo** [comment]
Any thoughts around maybe a fee limit on the list? Allow signing up to a fee limit.

> One important fact to remember is that any signer list that can sign transactions can drain the account's XRP via fees.

**Reply by shortthefomo** [reply]
eg. Another very useful one would be transaction size. This would possibly make exchanges use signer lists over a standard signature.

Eg. 1-200 XRP list a can sign, 200-2000 list b can sign… and so on requiring more signatures with size.

**Reply by mvadari** [reply]
That adds a lot of complexity and configuration. This spec (and the alternate one) are both very efficient to run within the current multisig framework. If you need to add amount checks, that adds a lot of overhead.

**Reply by WietseWind** [reply]
Sounds like a perfect use case for 🪝 😇

**Comment by tequdev** [comment]
Will the TransactionTypeBitmask and SignerListID correspond?

If so, is it correct that there can be only one list with the same TransactionTypeBitmask, and that when deleting a SignerList, the TransactionTypeBitmask is used as the key?

**Reply by tequdev** [reply]
It looks like multiple SignerLists are created when multiple transaction types are specified in the TransactionTypeBitmask field of a single SignerListSet transaction.

**Reply by mvadari** [reply]
For the existing spec, yes, multiple Signer lists are created/modified. This was suggested so that it is easier to set up a signer list that can authorize multiple transaction types.

For the Appendix B spec, the bitmask does not need to be unique. It is not included in the key. Multiple signer lists can have the same bitmask.

**Comment by ckeshava** [comment]
Regarding the calculation of the `SignerListID` field, I'm concerned about the potential collisions (earlier than the 256 transactions threshold). Many ledger objects use ASCII characters to compute the `key` field (Source: https://github.com/XRPLF/rippled/blob/ca3198164c2869406e1f04020f3c1a74c7d35485/src/ripple/protocol/impl/Indexes.cpp#L56)

For instance, this proposal could cause a collision between `LedgeNameSpace::BOOK_DIR` and a SignerList ledger object with a transaction type of value `65`. Is this correct?

I found it hard to test this, because `sha512Half` function takes an enum only.

**Reply by mvadari** [reply]
That is incorrect. The `key` field takes a prefix for the type of the object (which would be the value `65` for `BOOK_DIR`) and appends on the additional information. There cannot be collisions between objects of different types (assuming that you don't find a hash collision between two separate initial values, of course).

**Reply by ckeshava** [reply]
I see, okay. So `ltSIGNER_LIST` is the Space Key. This is used along with `accountAddress + (1 + TxType)` to get the final `SignerListID` 👍

**Reply by F171717** [reply]
> I see, okay. So `ltSIGNER_LIST` is the Space Key. This is used along with `accountAddress + (1 + TxType)` to get the final `SignerListID

**Comment by ckeshava** [comment]
I like that we can modify multiple transaction types in a single shot. Otherwise, it is difficult for the account administrators to set up a signer_list for every transaction type (assuming they don't want to fall back on the global signer_list)
